source/websocket.c

```diff
@@ -0,0 +1,908 @@
+
+    aws_linked_list_init(&websocket->thread_data.outgoing_frame_list);
+
+    aws_websocket_encoder_init(&websocket->thread_data.encoder, s_encoder_stream_outgoing_payload, websocket);
+
+    aws_websocket_decoder_init(&websocket->thread_data.decoder, s_decoder_on_frame, s_decoder_on_payload, websocket);
+
+    int err = aws_mutex_init(&websocket->synced_data.lock);
+    if (err) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_WEBSOCKET,
+            "static: Failed to initialize mutex, error %d (%s).",
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+
+        goto error;
+    }
+
+    aws_linked_list_init(&websocket->synced_data.outgoing_frame_list);
+
+    return &websocket->channel_handler;
+
+error:
+    websocket->channel_handler.vtable->destroy(&websocket->channel_handler);
+    return NULL;
+}
+
+static void s_handler_destroy(struct aws_channel_handler *handler) {
+    struct aws_websocket *websocket = handler->impl;
+    AWS_LOGF_TRACE(AWS_LS_HTTP_WEBSOCKET, "id=%p: Destroying websocket.", (void *)websocket);
+
+    aws_mutex_clean_up(&websocket->synced_data.lock);
+    aws_mem_release(websocket->alloc, websocket);
+}
+
+void aws_websocket_acquire_hold(struct aws_websocket *websocket) {
+    size_t prev_refcount = aws_atomic_fetch_add(&websocket->refcount, 1);
+    AWS_LOGF_TRACE(
+        AWS_LS_HTTP_WEBSOCKET,
+        "id=%p: Websocket refcount increased, currently %zu.",
+        (void *)websocket,
+        prev_refcount + 1);
+
+    if (prev_refcount == 0) {
+        /* Prevent channel from destroying the websocket unexpectedly */
+        aws_channel_acquire_hold(websocket->channel_slot->channel);
+    }
+}
+
+void aws_websocket_release_hold(struct aws_websocket *websocket) {
+    assert(websocket);
+    assert(websocket->channel_slot);
+
+    size_t prev_refcount = aws_atomic_fetch_sub(&websocket->refcount, 1);
+    if (prev_refcount == 1) {
+        AWS_LOGF_TRACE(
+            AWS_LS_HTTP_WEBSOCKET,
+            "id=%p: Final websocket refcount released, shut down if necessary.",
+            (void *)websocket);
+
+        /* Channel might already be shut down, but make sure */
+        aws_channel_shutdown(websocket->channel_slot->channel, AWS_ERROR_SUCCESS);
+
+        /* Channel won't destroy its slots/handlers until its refcount reaches 0 */
+        aws_channel_release_hold(websocket->channel_slot->channel);
+
+    } else {
+        assert(prev_refcount != 0);
+
+        AWS_LOGF_TRACE(
+            AWS_LS_HTTP_WEBSOCKET,
+            "id=%p: Websocket refcount released, %zu remaining.",
+            (void *)websocket,
+            prev_refcount - 1);
+    }
+}
+
+/* Insert frame into list, sorting by priority, then by age (high-priority and older frames towards the front) */
+static void s_enqueue_prioritized_frame(struct aws_linked_list *list, struct outgoing_frame *frame) {
```

Comment:
It's N, I assumed there wouldn't be many queued frames.

I'll try to hook it up with priority queue...

Reply:
There's a lot more error-handling involved when using priority-queues and array-lists.

I changed reversed the order we scan so that in the common case we insert at the first place we look